### ISSUE TITLE
one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [512]] is at version 3; expected version 2 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient

### ISSUE THREAD

**Issue by Here-Liu** [Lornatang/ESRGAN-PyTorch#5]
When the code run at '**scaler.scale(d_loss).backward()**' , I got an error："one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [512]] is at version 3; expected version 2 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True)."

It has been bothering me for many days, can someone help me?

**Comment by Lornatang** [comment]
I check it.

**Comment by yi-zhi-shu-shu** [comment]
> I check it.

Thanks a lot !!!!

**my code is same as yours with [torch 1.8.1 cuda 10.2]**

########################## **Train**

	# switch to train mode
	discriminator.train()
	generator.train()

	for _, train_data in enumerate(train_loader): 
		lr =train_data['lr'].cuda(args.gpu, non_blocking=True)
		hr =train_data['hr'].cuda(args.gpu, non_blocking=True)
		batch_size = lr.size(0)

		real_label = torch.full((batch_size, 1), 1, dtype=lr.dtype).cuda(args.gpu, non_blocking=True)
		fake_label = torch.full((batch_size, 1), 0, dtype=lr.dtype).cuda(args.gpu, non_blocking=True)

		##############################################
		# (1) Update D network: E(hr)[fake(C(D(hr) - E(sr)C(sr)))] + E(sr)[fake(C(fake) - E(real)C(real))]
		##############################################
		discriminator_optimizer.zero_grad()

		with amp.autocast():
			sr = generator(lr)
			# It makes the discriminator distinguish between real sample and fake sample.
			real_output = discriminator(hr)
			fake_output = discriminator(sr.detach())

			d_loss_real = adversarial_criterion(real_output - torch.mean(fake_output), real_label)
			d_loss_fake = adversarial_criterion(fake_output - torch.mean(real_output), fake_label)

			d_loss = d_loss_real + d_loss_fake

		scaler.scale(d_loss).backward()
		scaler.step(discriminator_optimizer)
		scaler.update()
                ...

####################################################

the "train_psnr" work fine, I think problem may in discriminator
